app/usedgoodreads/__main__.py

```diff
@@ -15,32 +15,59 @@ def index(key):
     cache = Book.get(key=key)
 
     if cache:
+        # Book is in database
         books = UsedBook.get(isbn=cache.isbn)
-
         if books:
+            # Used books are in database
             results = [{"title": v.title, "description": v.description,
                         "price": v.price, "link": v.link} for v in books]
 
             results.sort(key=lambda each: each["price"])
 
-            return render_template("index.html", isbn=cache.isbn, title=cache.title,
+            return render_template("index.html", isbn=cache.isbn,
+                                   title=cache.title,
                                    results=results)
-
-    ticket = Ticket(key)
-
-    if q.fetch_job(ticket.isbn_jid):
-        return render_template("index.html")
-
-    if q.fetch_job(ticket.used_books_jid):
-        return render_template("index.html", status="searching")  # TODO: Should return actual status
-
-    q.enqueue(resolve_goodreads_key, ticket, job_timeout=60 * 1,
-            ttl=60 * 60, job_id=ticket.isbn_jid)
-
-    q.enqueue(resolve_used_books, key, job_timeout=60 * 1,
-              ttl=60 * 60, job_id=ticket.used_books_jid, depends_on=ticket.isbn_jid)
-
-    return render_template("index.html")
+        else:
+            # Used books need to be fetched
+            ticket = Ticket(key)
+            job = q.fetch_job(ticket.used_books_jid)
+            if job:
+                # Fetch job already in q
+                if job.get_status() == "finished":
+                    # There have been no results at the end of the fetch job
+                    return render_template("index.html",
+                                           isbn=cache.isbn,
+                                           title=cache.title,
+                                           status="Used books: No results")
+                else:
+                    # fetch job still running or there was an error
+                    return render_template("index.html",
+                                           isbn=cache.isbn,
+                                           title=cache.title,
+                                           status=f"Used books: {job.get_status()}")
+            else:
+                # Add fetch job to q
+                q.enqueue(resolve_used_books, key, job_timeout=60 * 1,
+                          ttl=60 * 60, job_id=ticket.used_books_jid)
+                return render_template("index.html",
+                                       isbn=cache.isbn,
+                                       title=cache.title,
+                                       status="Used books: enqueued")
+    else:
+        # Book is not in database
+        ticket = Ticket(key)
+        job = q.fetch_job(ticket.isbn_jid)
+        if job:
```

Comment:
Do we need the same logic here with checking for status finished etc.?

Reply:
No. Because if we can resolve the isbn, there will be an entry in the database for sure.
That is not the case for used books. It might be that we did resolve correctly but did not write anything to the db because there were no results.

Reply:
Ah, gotcha! Hm, then we are using two different places for persietence & storage:
1. the postgres database
2. the redis queue and its jobs

Should we only have persistence in postgres and see redis/rq only as a queue for workers?

Reply:
Decided on keeping it as is currently: #27